### DISCUSSION TITLE
what's the correct place to save user info in database when user signs in for the first time?

### DISCUSSION THREAD

**Discussion by stardvst** [nextauthjs/next-auth#3192] [Help] [answered]
when the user signs in for the first time (basically, sign up), i want to store user info (such as user's followings) in database.
obviously i want to do this only once.
what's the correct place in code to save user info in database?

**Comment by balazsorban44** [comment, marked as answer]
https://next-auth.js.org/adapters/models#user

**Reply by stardvst** [reply]
does this mean i cannot insert extra data while user is inserted into database?

**Reply by balazsorban44** [reply]
It means you absolutely can :smile:

**Reply by stardvst** [reply]
where? :))) in profile()? i have checked it, and it returns other profile information, not the "followings" which i need.

**Reply by balazsorban44** [reply]
> You can add more profile data by returning extra fields in your OAuth provider's profile() callback.

Make sure you pass the correct scopes to receive additional profile information from your provider.

Could you paste in your provider config?

**Reply by stardvst** [reply]
here it is:
```
  providers: [
    SpotifyProvider({
      clientId: process.env.SPOTIFY_CLIENT_ID,
      clientSecret: process.env.SPOTIFY_CLIENT_SECRET,
      authorization: {
        params: {
          scope: 'user-read-email user-follow-read',
        },
      },
    }),
  ],
```
in fact, i have provided the `user-follow-read` scope, and later i hit the `https://api.spotify.com/v1/me/following?type=artist&limit=50` endpoint in profile page for getting the followings and showing them in my profile page.

now i want to modify my code to store the followings in database once (instead of retrieving them in profile page), and read them from database and show in profile page.

**Reply by balazsorban44** [reply]
You can just move the API request into the `profile` callback, and return the result. It will be saved in the database.

Although remember, you might occasionally want to update the following number in the database anyway, otherwise, it will become out of sync.

**Reply by stardvst** [reply]
i need access_token to make the api request, which i think isn't available in profile().

i'm planning to do the sync part in a cron-job-like fashion. but i haven't researched that part well yet. if you have a suggestion where i should start looking, please let me know.

**Reply by balazsorban44** [reply]
You have the `access_token` :) https://next-auth.js.org/configuration/providers/oauth#options

**Reply by stardvst** [reply]
it's in tokens, okay thanks!

**Reply by stardvst** [reply]
@balazsorban44, if i'm storing `followings: Artist[]` in `profile` object, the profile object is going to be stored in database as is, yet it's not possible to store the array in database:
```
profile(profile, tokens) {
  return {
    id: profile.id,
    name: profile.display_name,
    email: profile.email,
    image: profile.images?.[0]?.url,
    followings: [{ name: 'artist', image: 'data;;' }],
  };
},
```
the error:
```
  4 function PrismaAdapter(p) {
  5  return {
→ 6   createUser: (data) => p.user.create({
        data: {
          name: 'name',
          email: 'email',
          image: undefined,
          followings: [
          ~~~~~~~~~~
            {
              name: 'artist',
              image: 'data;;'
            }
          ],
          emailVerified: null
        }
      })

Unknown arg `followings` in data.followings for type UserUncheckedCreateInput.
```

here are my models:
```
model User {
  id            String    @id @default(cuid())
  name          String?
  email         String?   @unique
  emailVerified DateTime?
  image         String?
  accounts      Account[]
  sessions      Session[]
  followings    Artist[]
}

model Artist {
  id    String  @id @default(cuid())
  name  String
  image String?
  users User[]

  @@unique([name])
}
```
and the explicit many-to-many relationship is modeled through a new table `_ArtistToUser`, into which i need to insert the followings data.

how should i store the followings in profile, and in database?

**Reply by balazsorban44** [reply]
In your model, that's not an array, just a relation.

You might have to create the artist with a Prisma query and connect it to the user in the profile callback.


or you could override the createUser adapter method and do it in one query https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#connect-or-create-a-record

I still think this will get out of sync pretty quickly. you basically want to keep around a copy of the Spotify API. 🤔

**Reply by stardvst** [reply]
okay, so the main reason i want to store the artist info in db is spotify rate limit. i need to get all users all following artists, then perform another bunch of api requests. so better if i at least have artist ids in db.
for my app i only need to have artist's spotify id, so i can store artist ids per user, and changed user the model to this:
```
model User {
  id            String    @id @default(cuid())
  name          String?
  email         String?   @unique
  emailVerified DateTime?
  image         String?
  accounts      Account[]
  sessions      Session[]
  followings    String[]
}
```
and artist model now becomes:
```
model Artist {
  id        String  @id @default(cuid())
  name      String
  image     String?
  spotifyId String

  @@unique([name])
}
```
however this doesn't store the 2 items in user table, user.followings is empty:
```
profile(profile, tokens) {
  return {
    id: profile.id,
    name: profile.display_name,
    email: profile.email,
    image: profile.images?.[0]?.url,
    followings: ['artist1_id', 'artist2_id'],
  };
},
```
